kano_profile/challenge_progress_tracker.py

```diff
@@ -19,13 +19,45 @@ class ChallengeProgressTracker(object):
     COMPLETED = 'completed'
     ABANDONED = 'abandoned'
 
+    _singleton_instance = None
+
+    @staticmethod
+    def get_instance():
+        if not ChallengeProgressTracker._singleton_instance:
+            ChallengeProgressTracker()
+
+        return ChallengeProgressTracker._singleton_instance
+
     def __init__(self):
+        if ChallengeProgressTracker._singleton_instance:
+            raise Exception('This class is a singleton!')
+        else:
+            ChallengeProgressTracker._singleton_instance = self
+
+        self._app_name = ''
         self._tracking_mode = None
         self._chal_group = None
         self._chal_no = None
         self._chal_outcome = None
         self._app_session_id = str(uuid4())
 
+    def set_name(self, app_name):
+        if not isinstance(app_name, basestring):
+            raise TypeError('set_name argument must be of type string')
+
+        if app_name:
+            self._app_name = app_name
+        else:
+            raise ValueError('App name must not be empty')
+
+    def _check_name_is_set(self):
```

Comment:
Given that we run this at the start of every function, it might be nice to make this a decorator. It is fine as it is if it is too much effort.